Error code E0794 for late-bound lifetime parameter error. (#107416)
rust-lang/rust: compiler/rustc_error_codes/src/error_codes/E0793.md

A lifetime parameter of a function definition is called *late-bound* if it both:

1. appears in an argument type
2. does not appear in a generic type constraint

You cannot substitute concrete lifetimes for late-bound lifetime parameters.

Erroneous code example:

> ```compile_fail,E0793

Reviewer: Please add an empty line before code blocks (below too).

A lifetime parameter of a function definition is called *late-bound* if it both:

> 1. appears in an argument type

Reviewer: Please try to follow the formatting of error code:

```
[small explanation (one/two sentence)]

Erroneous code example:

[erroneous code example]

[long explanation]

[fixed code example]
```

Author: @GuillaumeGomez
I refactored it now. Is the last bit of information OK, or should I merge it in [long explanation]?

Reviewer: It's fine like this, thanks.

> 2. does not appear in a where clause

Reviewer 2: I think it would be better to talk about not appearing in "a generic type constraint" rather than in "a where clause", as the `'a` lifetime parameters in both `fn<'a, T: Trait<'a>>(t: T)` and `fn<'a>(t: impl Trait<'a>)` are early bound without there being any `where` clause.

Author: Good point!

Function types are not generic over late-bound lifetime parameters,
but are universally quantified over them. This means that when referring to a

> function of a concrete type by substituting concrete types for
> generic ones, you cannot specify a late-bound lifetime, because such lifetime
> parameters are resolved at the function's call site(s).

Reviewer 2: ```suggestion
concrete instance of the generic function, you cannot specify any of
its late-bound lifetime parameters, because they are not resolved until
each call site(s).
```

Erroneous code example:
```compile_fail,E0793
fn foo<'a>(x: &'a str) -> &'a str { x }
let _ = foo::<'static>;
```

To fix the issue, remove the specified late-bound lifetime parameter:
```
fn foo<'a>(x: &'a str) -> &'a str { x }
let _ = foo;
```

Lifetime parameters that are not late-bound are called *early-bound*.

> The confusion might stem from the fact that late-bound and early-bound

Reviewer 2: ```suggestion
Confusion may arise from the fact that late-bound and early-bound
```

lifetime parameters are specified the same way in function definitions.
When referring to a function pointer type, universal quantification over
late-bound lifetime parameters can be made explicit:
```
trait BarTrait<'a> {}

struct Bar<'a> {
    s: &'a str
}

impl<'a> BarTrait<'a> for Bar<'a> {}

fn bar<'a, 'b, T>(x: &'a str, t: T) -> &'a str
where T: BarTrait<'b>
{
    x
}

let bar_fn: for<'a> fn(&'a str, Bar<'static>) -> &'a str = bar;
```
In the definition of `bar`, the lifetime parameter `'a` is late-bound, while
`'b` is early-bound. This is reflected in the type annotation for `bar_fn`,
where `'a` is universally quantified and `'b` is substituted by a specific

> lifetime.

Reviewer 2: ```suggestion
lifetime.  It is not possible to explicitly specify early-bound lifetime
parameters when late-bound parameters are present, although the types
that are constrained by early-bound parameters can be specified (as in the
example above).
```

Author: Is this what you mean? https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=c30a35a698a82969827f8260139c0f9d
It is possible in the sense that it just generates a warning, but warns about it being an error in the future.

Reviewer 2: Apologies, only just saw your comment (the trials and tribulations of GitHub mobile!).  Yes, `_bar_fn2` in your playground is what I meant—a number of the tests output the error message `"cannot specify lifetime arguments explicitly if late bound lifetime parameters are present"`, but I didn't appreciate they were only warnings (which compiletest denies/upgrades to errors); I think it'd still be helpful to make some reference to it?

That `_bar_fn3` in your playground compiles is completely consistent with your existing explanation: no attempt was made to (erroneously) specify concrete lifetimes for late-bound parameters in a function reference, nor was the function reference cast to a pointer that would require explicit universal quantification; lifetime parameter `'b` is early-bound to the lifetime parameter of `Bar`, whatever that ends up happening to be—it is not universally quantified.
